### PR TITLE
Fix contrast limit slider for small range

### LINKED ISSUE #3893: contrast limits slider broken with very small max value
## 🐛 Bug

I was working with an image today with a very small max value, 2.6e-17, but a lot of richness around 10-20 etc, and found that the contrast limits slider is not adjustable, i.e. when i click on it I can't drag it around.

## To Reproduce

```python
v = napari.Viewer()
v.add_image(np.random.random((10, 10))*2.6e-17)
```

## Expected behavior

I should be able to adjust contrast limits value

## Environment

Current main, and latest superqt

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
Closes #3893 by fixing the contrast limits slider for small range values.

@tlambert03 can you review this - let me know if you can imagine any problems here? Also any suggested improvements to doc strings or location of added function are welcome.

I didn't add any tests yet - hard to maybe add tests for whole behaviour, but any suggestions welcome here too

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# References
Closes #3893

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/_qt/layer_controls/qt_image_controls_base.py [hunk 1]
```diff
@@ -259,3 +262,33 @@ def reset():
             range_btn.setFixedWidth(65)
             range_btn.clicked.connect(layer.reset_contrast_limits_range)
             self._layout.addWidget(range_btn, alignment=Qt.AlignBottom)
+
+
+def range_to_decimals(range_, dtype):
```
Comment (tlambert03) [inline napari/_qt/layer_controls/qt_image_controls_base.py:256]: this is fine, but I assume you meant to delete this code from QContrastLimitsPopup above?  currently it's duplicated

**Review by tlambert03 — CHANGES_REQUESTED** [review]

File: napari/_qt/layer_controls/qt_image_controls_base.py (hunk 1 above)
Comment (sofroniewn) [reply napari/_qt/layer_controls/qt_image_controls_base.py:256]: lol yes - and replace it there too. will fix - thanks!!

Comment (sofroniewn) [reply napari/_qt/layer_controls/qt_image_controls_base.py:256]: done

**Review by tlambert03 — APPROVED** [review]
LGTM!

**Comment by sofroniewn** [comment]
This is actually failing because of

```python-traceback
  napari/_qt/layer_controls/qt_image_controls_base.py:273: in range_to_decimals
      if np.issubdtype(dtype, np.integer):
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  arg1 = dtype("float64"), arg2 = <class 'numpy.integer'>
  
      @set_module('numpy')
      def issubdtype(arg1, arg2):
          r"""
          Returns True if first argument is a typecode lower/equal in type hierarchy.
      
          This is like the builtin :func:`issubclass`, but for `dtype`\ s.
      
          Parameters
          ----------
          arg1, arg2 : dtype_like
              `dtype` or object coercible to one
      
          Returns
          -------
          out : bool
      
          See Also
          --------
          :ref:`arrays.scalars` : Overview of the numpy type hierarchy.
          issubsctype, issubclass_
      
          Examples
          --------
          `issubdtype` can be used to check the type of arrays:
      
          >>> ints = np.array([1, 2, 3], dtype=np.int32)
          >>> np.issubdtype(ints.dtype, np.integer)
          True
          >>> np.issubdtype(ints.dtype, np.floating)
          False
      
          >>> floats = np.array([1, 2, 3], dtype=np.float32)
          >>> np.issubdtype(floats.dtype, np.integer)
          False
          >>> np.issubdtype(floats.dtype, np.floating)
          True
      
          Similar types of different sizes are not subdtypes of each other:
      
          >>> np.issubdtype(np.float64, np.float32)
          False
          >>> np.issubdtype(np.float32, np.float64)
          False
      
          but both are subtypes of `floating`:
      
          >>> np.issubdtype(np.float64, np.floating)
          True
          >>> np.issubdtype(np.float32, np.floating)
          True
      
          For convenience, dtype-like objects are allowed too:
      
          >>> np.issubdtype('S1', np.string_)
          True
          >>> np.issubdtype('i4', np.signedinteger)
          True
      
          """
          if not issubclass_(arg1, generic):
  >           arg1 = dtype(arg1).type
  E           TypeError: Cannot interpret 'dtype("float64")' as a data type
```

Does @tlambert03 @grlee77 @jni have any suggestion on how to fix this? We could coerce or avoid the call? We're just checking if the data are integers I believe

**Comment by tlambert03** [comment]
do you understand why this is failing now, with the `range_to_decimals` where it is, but it wasn't before?

**Comment by sofroniewn** [comment]
> do you understand why this is failing now, with the range_to_decimals where it is, but it wasn't before?

I think before code with `np.issubdtype` was only called in the `QContrastLimitsPopup` which wasn't tested with the full range of dtypes that the main image layer gets tested with.

**Comment by sofroniewn** [comment]
I should add the `np.issubdtype` call is only failing for Tensorstore arrays, so maybe this needs to be fixed on their end or a special patch is needed. @jni can you take a look? I'm not so familiar with tensorstore

**Comment by sofroniewn** [comment]
Just bumping this, maybe @grlee77 or @Carreau have some recommendations on how to deal with the `np.issubdtype` call failing for Tensorstore arrays. Not sure what we can do or if we might need to fix something for tensorstore, cc @jni. We can reach out to their devs if need be too, but I thought we should try and fix it ourselves first

**Comment by grlee77** [comment]
I can try to take a look now (pretty familiar with NumPy dtypes, but not `tensorstore` itself)

File: napari/_qt/layer_controls/qt_image_controls_base.py [hunk 2]
```diff
@@ -259,3 +251,33 @@ def reset():
...
+        Data type of the layer. Integers layers are given integer.
+        step sizes.
+
+    Returns
+    -------
+    int
+        Decimals of precision.
+    """
+
+    if np.issubdtype(dtype, np.integer):
```
Comment (grlee77) [inline napari/_qt/layer_controls/qt_image_controls_base.py:277]: The following works to support `tensorstore.dtype` objects  [tensorstore API page here](https://google.github.io/tensorstore/python/api/tensorstore.dtype.html#tensorstore.dtype))

```suggestion
 
    if hasattr(dtype, 'numpy_dtype'):
        # retrieve the corresponding numpy.dtype from a tensorstore.dtype
        dtype = dtype.numpy_dtype

    if np.issubdtype(dtype, np.integer):
```

**Review by grlee77 — DISMISSED** [review]

**Comment by sofroniewn** [comment]
> I can try to take a look now (pretty familiar with NumPy dtypes, but not tensorstore itself)

Nice! Your suggestion looks great to me, thanks! I've committed. If all tests pass then I think this is ready for re-review/ merge

**Comment by grlee77** [comment]
There are failures now, but different ones: https://github.com/napari/napari/runs/4796393301?check_suite_focus=true#step:9:272

It looks like the expected number of decimals and comments in `test_clim_slider_step_size_and_precision` (in `_qt/layer_controls/_tests/test_qt_image_base_layer_.py`) needs to be updated to match the values returned by `range_to_decimals`

**Comment by grlee77** [comment]
Here is a patch that fixes the failures. I can make a PR to your branch if that is easier

```
diff --git a/napari/_qt/layer_controls/_tests/test_qt_image_base_layer_.py b/napari/_qt/layer_controls/_tests/test_qt_image_base_layer_.py
index 6d8a02f2..40a93d43 100644
--- a/napari/_qt/layer_controls/_tests/test_qt_image_base_layer_.py
+++ b/napari/_qt/layer_controls/_tests/test_qt_image_base_layer_.py
@@ -10,6 +10,7 @@ from napari._qt.layer_controls.qt_image_controls_base import (
     QContrastLimitsPopup,
     QRangeSliderPopup,
     QtBaseImageControls,
+    range_to_decimals,
 )
 from napari.layers import Image, Surface
 
@@ -101,10 +102,13 @@ def test_clim_slider_step_size_and_precision(qtbot, mag):
     popup = QContrastLimitsPopup(layer)
     qtbot.addWidget(popup)
 
-    # the range slider popup labels should have a number of decimal points that
-    # is inversely proportional to the order of magnitude of the range of data,
-    # but should never be greater than 5 or less than 0
-    decimals = min(6, max(int(3 - mag), 0))
+    # scale precision with the log of the data range order of magnitude
+    # eg.   0 - 1   (0 order of mag)  -> 3 decimal places
+    #       0 - 10  (1 order of mag)  -> 2 decimals
+    #       0 - 100 (2 orders of mag) -> 1 decimal
+    #       ≥ 3 orders of mag -> no decimals
+    # no more than 64 decimals
+    decimals = range_to_decimals(layer.contrast_limits, layer.dtype)
     assert popup.slider.decimals() == decimals
 
     # the slider step size should also be inversely proportional to the data
```

**Comment by sofroniewn** [comment]
@grlee77 if you can PR to my branch or if you have permissions just push directly to this PR then that would be great! thanks for chasing down fix!!

**Comment by sofroniewn** [comment]
Actually @tlambert03 do you see any concerns here, as the following will no longer be true?

>  the range slider popup labels should have a number of decimal points that
  is inversely proportional to the order of magnitude of the range of data,
  but should never be greater than 5 or less than 0

**Comment by tlambert03** [comment]
what's the case now?  is the number of decimals unbound?  that's definitely problematic for guis.  something like 11 decimals (like the test seems to indicate) does seem like it would create an ugly UI.  can you confirm whether that happens here?

_edit_: I know we've discussed this elsewhere... it kinda seems like we need to switch to scientific notation, which of course is more work :/

**Comment by grlee77** [comment]
> what's the case now? is the number of decimals unbound? that's definitely problematic for guis. something like 11 decimals (like the test seems to indicate) does seem like it would create an ugly UI. can you confirm whether that happens here?

It is bound to 64 decimal places in this PR which is not reasonable for display. Here is what it looks like for data in the range (-1e-16, 3e-16) and the values are not visible after 14 decimals. Perhaps we can change the visualization to use scientific notation beyond 5 digits or so.

clipped

e.g. running the following in the console to get a 2D image with a very narrow range

```Python
import numpy as np

small_range_image = np.linspace(-1e-16, 3e-16, 64*64).reshape(64, 64)
viewer.add_image(small_range_image)

```

**Comment by tlambert03** [comment]
> the values are not visible after 14 decimals. Perhaps we can change the visualization to use scientific notation beyond 5 digits or so.

thanks for taking a look @grlee77.  I'm fine with merging this as is, and working on the scientific notation in a follow up.  S'ok with me to just comment out the part of the test that checked the decimals on the UI

**Comment by grlee77** [comment]
According to the Qt docs on [QDoubleSpinBox](https://doc.qt.io/qt-5/qdoublespinbox.html) there is a hard limit of 18 characters that can be displayed. Do you think it is worth playing with using the `validate()`, `textFromValue()` and `valueFromText()` to customize the display? It looks like that code would go in `superqt` itself rather than napari. 

If so, we can do that in a followup so that this can be merged (it just needs sofroniewn/napari#11 merged so the test fix shows up here)

**Comment by tlambert03** [comment]
Sure text from value is likely to work. It could go here as well, just a thin subclass for this purpose (but, I think an upstream PR would also be handy since it's almost always going to be useful right?)

**Comment by sofroniewn** [comment]
I have now merged the patch, so this should be ready to merge once all tests pass! Thanks all

**Pull request merged**